internal/api/errorcodes.go

```diff
@@ -0,0 +1,76 @@
+	ErrorCodeTooManyEnrolledMFAFactors         ErrorCode = "too_many_enrolled_mfa_factors"
+	ErrorCodeMFAFactorNameConflict             ErrorCode = "mfa_factor_name_conflict"
+	ErrorCodeMFAFactorNotFound                 ErrorCode = "mfa_factor_not_found"
+	ErrorCodeMFAIPAddressMismatch              ErrorCode = "mfa_ip_address_mismatch"
+	ErrorCodeMFAChallengeExpired               ErrorCode = "mfa_challenge_expired"
+	ErrorCodeMFAVerificationFailed             ErrorCode = "mfa_verification_failed"
+	ErrorCodeMFAVerificationRejected           ErrorCode = "mfa_verification_rejected"
+	ErrorCodeInsufficientAAL                   ErrorCode = "insufficient_aal"
+	ErrorCodeCaptchaFailed                     ErrorCode = "captcha_failed"
+	ErrorCodeSAMLProviderDisabled              ErrorCode = "saml_provider_disabled"
+	ErrorCodeManualLinkingDisabled             ErrorCode = "manual_linking_disabled"
+	ErrorCodeSMSSendFailed                     ErrorCode = "sms_send_failed"
+	ErrorCodeEmailNotConfirmed                 ErrorCode = "email_not_confirmed"
+	ErrorCodePhoneNotConfirmed                 ErrorCode = "phone_not_confirmed"
+	ErrorCodeReauthNonceMissing                ErrorCode = "reauth_nonce_missing"
+	ErrorCodeSAMLRelayStateNotFound            ErrorCode = "saml_relay_state_not_found"
+	ErrorCodeSAMLRelayStateExpired             ErrorCode = "saml_relay_state_expired"
+	ErrorCodeSAMLIdPNotFound                   ErrorCode = "saml_idp_not_found"
+	ErrorCodeSAMLAssertionNoUserID             ErrorCode = "saml_assertion_no_user_id"
+	ErrorCodeSAMLAssertionNoEmail              ErrorCode = "saml_assertion_no_email"
+	ErrorCodeUserAlreadyExists                 ErrorCode = "user_already_exists"
+	ErrorCodeSSOProviderNotFound               ErrorCode = "sso_provider_not_found"
+	ErrorCodeSAMLMetadataFetchFailed           ErrorCode = "saml_metadata_fetch_failed"
+	ErrorCodeSAMLIdPAlreadyExists              ErrorCode = "saml_idp_already_exists"
+	ErrorCodeSSODomainAlreadyExists            ErrorCode = "sso_domain_already_exists"
+	ErrorCodeSAMLEntityIDMismatch              ErrorCode = "saml_entity_id_mismatch"
+	ErrorCodeConflict                          ErrorCode = "conflict"
+	ErrorCodeProviderDisabled                  ErrorCode = "provider_disabled"
+	ErrorCodeUserSSOManaged                    ErrorCode = "user_sso_managed"
+	ErrorCodeReauthenticationNeeded            ErrorCode = "reauthentication_needed"
+	ErrorCodeSamePassword                      ErrorCode = "same_password"
+	ErrorCodeReauthenticationNotValid          ErrorCode = "reauthentication_not_valid"
+	ErrorCodeOTPExpired                        ErrorCode = "otp_expired"
+	ErrorCodeOTPDisabled                       ErrorCode = "otp_disabled"
+	ErrorCodeIdentityNotFound                  ErrorCode = "identity_not_found"
+	ErrorCodeWeakPassword                      ErrorCode = "weak_password"
+	ErrorCodeOverRequestRate                   ErrorCode = "over_request_rate"
```

Comment:
maybe `ErrorCodeRateLimitExceeded` and `rate_limit_exceeded` ?

Reply:
Wanted to keep it closer to the email and SMS ones.

Reply:
Coming back to this, I still feel like `over_request_rate`, `over_email_send_rate` and `over_sms_send_rate` more succinctly express the issue at hand. Rate limit exceeded is not very specific ("what rate limit").

I'll add a `_limit` suffix to make it clearer though.